R/matrix_utils.R

```diff
@@ -0,0 +1,70 @@
+# matrix_utils.R: utilities for matrix conversion
+#
+# This file defines the to_matrix() function, which can be used to convert
+# data.frame or other types of matrix-like objects to matrix for use in
+# mlpack bindings.
+#
+# mlpack is free software; you may redistribute it and/or modify it under the
+# terms of the 3-clause BSD license.  You should have received a copy of the
+# 3-clause BSD license along with mlpack.  If not, see
+# http://www.opensource.org/licenses/BSD-3-Clause for more information.
+
+# Given some matrix-like X (which should be either a matrix or
+# data.frame), convert it into a matrix.
+to_matrix <- function(x) {
+  if (!is.matrix(x) && !is.data.frame(x)) {
+    stop("given argument is not matrix-like")
+  }
+  if (is.matrix(x)) {
+    return(x)
+  }
+  else if (is.data.frame(x)) {
+    y <- data.matrix(x)
+    return(y)
+  }
+}
+
+# Given some matrix-like X (which should be either a matrix or
+# data.frame), convert it into a matrix.
+to_matrix_with_info <- function(x, points_are_rows) {
+  if (!is.matrix(x) && !is.data.frame(x)) {
+    stop("given argument is not matrix-like")
+  }
+  # It is already a matrix, so the vector of info is all 0s (all numeric).
+  if (is.matrix(x)) {
+    if (points_are_rows) {
+      d <- matrix(FALSE, ncol = dim(x)[2])
+    } else {
+      d <- matrix(FALSE, nrow = dim(x)[1])
+    }
+    return(list("info" = d, "data" = x))
+  }
+  else if (is.data.frame(x)) {
+    if (points_are_rows) {
+      # Construct dataset information: 1s represent categorical data, 0s
+      # represent otherwise.
+      d <- matrix(FALSE, ncol = dim(x)[2])
+      char_data <- which(sapply(x, class) %in% c("factor", "character"))
+      if (length(char_data) != 0) {
+        for (i in 1:length(char_data)) {
+          d[char_data[i]] <- TRUE
+          x[, char_data[i]] <- as.numeric(as.factor(unlist(x[, char_data[i]])))
+        }
+      }
+    }
+    else {
+      d <- matrix(FALSE, nrow = dim(x)[1])
+      options(warn = -1)
+      char_data <- which(is.na(as.character(as.numeric(x[, 1]))))
+      options(warn = 0)
+      if (length(char_data) != 0) {
+        for (i in 1:length(char_data)) {
+          d[char_data[i]] <- TRUE
+          x[char_data[i], ] <- as.numeric(as.factor(unlist(x[char_data[i], ])))
+        }
+      }
+    }
+    y <- data.matrix(x)
```

Comment:
Hi I think so test is failing here. 
Can we disable this part when `points_are_rows=F`, and can show warning message if user pass`points_are_rows=F` with `matrix_with_info` as : `points_are_rows` will always treated as `TRUE` in case of `data.frame`.?

As this will results into` to_matrix_with_info()` will only accept this type of data.frame: 
``` 
   a  b  c  d  e  f
1  1  2  3  4  5  a
2  6  7  8  9 10  b
3 11 12 13 14 15  c
```
and not this type of data.frame:
```
   a  b  c  d  e
1  1  2  3  4  5
2  6  7  8  9 10
3 11 12 13 14 15
4  a  b  c  d  e 
```
Thank You.

Reply:
Agreed.  See below.  Of these two only the first is sensible in R where _a data.frame is a list of column vectors each of which contains **one** and only one type of data_.

Reply:
I have updated this part.